x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportSetUpgradeModeAction.java

```diff
@@ -0,0 +1,327 @@
+        if (isRunning.compareAndSet(false, true) == false) {
+            String msg = "Attempted to set [upgrade_mode] to [" +
+                request.isEnabled() + "] from [" + MlMetadata.getMlMetadata(state).isUpgradeMode() +
+                "] while previous request was processing.";
+            Exception detail = new IllegalStateException(msg);
+            listener.onFailure(new ElasticsearchStatusException(
+                "Cannot change [upgrade_mode]. Previous request is still being processed.",
+                RestStatus.TOO_MANY_REQUESTS,
+                detail));
+            return;
+        }
+
+        // Noop, nothing for us to do, simply return fast to the caller
+        if (request.isEnabled() == MlMetadata.getMlMetadata(state).isUpgradeMode()) {
+            isRunning.set(false);
+            listener.onResponse(new AcknowledgedResponse(true));
+            return;
+        }
+
+        ActionListener<AcknowledgedResponse> wrappedListener = ActionListener.wrap(
+            r -> {
+                isRunning.set(false);
+                listener.onResponse(r);
+            },
+            e -> {
+                isRunning.set(false);
+                listener.onFailure(e);
+            }
+        );
+        final PersistentTasksCustomMetaData tasksCustomMetaData = state.metaData().custom(PersistentTasksCustomMetaData.TYPE);
+
+        // <4> We have unassigned the tasks, respond to the listener.
+        ActionListener<List<PersistentTask<?>>> unassignPersistentTasksListener = ActionListener.wrap(
+            unassigndPersistentTasks -> {
+                // Wait for our tasks to all stop
+                awaitCondition(
+                    () -> client.admin()
+                        .cluster()
+                        .prepareListTasks()
+                        .setActions(DATAFEED_TASK_NAME + "[c]", JOB_TASK_NAME + "[c]")
+                        .get()
+                        .getTasks()
+                        .isEmpty(),
+                    "datafeed and job tasks to fully stop.",
+                    request.timeout(),
+                    wrappedListener);
+            },
+            wrappedListener::onFailure
+        );
+
+        // <3> After isolating the datafeeds, unassign the tasks
+        ActionListener<List<IsolateDatafeedAction.Response>> isolateDatafeedListener = ActionListener.wrap(
+            isolatedDatafeeds -> unassignPersistentTasks(tasksCustomMetaData, unassignPersistentTasksListener),
+            wrappedListener::onFailure
+        );
+
+        /*
+          <2> Handle the cluster response and act accordingly
+          <.1>
+              If we are enabling the option, we need to isolate the datafeeds so we can unassign the ML Jobs
+          </.1>
+          <.2>
+              If we are disabling the option, we need to wait to make sure all the jobs get reallocated to an appropriate node
```

Comment:
@droberts195 yeah, I have been thinking about this. The conditions on return have been a constant internal debate. 

Waiting for an `getAssignmentExplanation().equals(AWAITING_UPGRADE.explanation) == false` could work.

Reply:
@droberts195 that simple check works for the job tasks. However, for the datafeed tasks, I had to make an additional check to verify that the job assignmentId has converged. 

This *SHOULD* happen because when the new assignment and the old one are equivalent, that is a no-op in the reassignment logic